python/langsmith/wrappers/_openai.py

```diff
@@ -0,0 +1,90 @@
+from __future__ import annotations
+
+import functools
+from typing import TYPE_CHECKING, Callable, List, TypeVar, Union
+
+from langsmith import run_helpers
+
+if TYPE_CHECKING:
+    from openai import AsyncOpenAI, OpenAI
+    from openai.types.chat.chat_completion_chunk import (
+        ChatCompletionChunk,
+    )
+    from openai.types.completion import Completion
+
+C = TypeVar("C", bound=Union["OpenAI", "AsyncOpenAI"])
+
+
+def _reduce_chat(all_chunks: List[ChatCompletionChunk]) -> dict:
+    all_content = []
+    for chunk in all_chunks:
+        content = chunk.choices[0].delta.content
+        if content is not None:
+            all_content.append(content)
+    content = "".join(all_content)
+    if all_chunks:
+        d = all_chunks[-1].model_dump()
+        d["choices"] = [{"message": {"role": "assistant", "content": content}}]
+    else:
+        d = {"choices": [{"message": {"role": "assistant", "content": content}}]}
+
+    return d
+
+
+def _reduce_completions(all_chunks: List[Completion]) -> dict:
+    all_content = []
+    for chunk in all_chunks:
+        content = chunk.choices[0].text
+        if content is not None:
+            all_content.append(content)
+    content = "".join(all_content)
+    if all_chunks:
+        d = all_chunks[-1].model_dump()
+        d["choices"] = [{"text": content}]
+    else:
+        d = {"choices": [{"text": content}]}
+
+    return d
+
+
+def _get_wrapper(original_create: Callable, name: str, reduce_fn: Callable) -> Callable:
+    @functools.wraps(original_create)
+    def create(*args, stream: bool = False, **kwargs):
+        decorator = run_helpers.traceable(
+            name=name, run_type="llm", reduce_fn=reduce_fn if stream else None
+        )
+
+        return decorator(original_create)(*args, stream=stream, **kwargs)
+
+    @functools.wraps(original_create)
+    async def acreate(*args, stream: bool = False, **kwargs):
```

Comment:
Could probably simplify this. Main reason for it is that the response type may or may not be a generator, and we don't handle that well by default